pci/src/vfio.rs

```diff
@@ -1059,16 +1055,18 @@ impl VfioPciDevice {
     }
 
     fn generate_user_memory_regions<F>(
-        region_index: u32,
-        region_start: u64,
-        region_size: u64,
-        host_addr: u64,
+        device: &Arc<VfioDevice>,
+        region: &MmioRegion,
```

Comment:
I need to think about the big picture for the VFIO case and assess if your approach is appropriate here. I'm thinking about integrating the support for sparse regions for VFIO (which is already supported as part of `vfio-ioctls`) and I want to make sure your code can help with that.

Reply:
As we discussed offline, I reverted the changes on `vfio.rs` from my last commit so that @sboeuf can add the sparse region support for VFIO devices while this refactoring. As a reference, here is my changes on `vfio.rs`: https://github.com/likebreath/cloud-hypervisor/commit/6cb0ee1d3ef25bd78cc5d570a01f7b3c76ee5fb4.